[custom_op] explicit autograd API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #101824
* #101823
* #101822

This PR adds an explicit API for registering a backward formula for a
CustomOp. In the end state, we will likely have this explicit API and a
magic API (which is sugar on top of an explicit API), since different
parties of users prefer different ones.

Concretely, to define a backward formula for a CustomOp:
- a user must provide us a "save for backward" function that accepts
(inputs, output) and returns exactly what they want saved for backward
- a user must provide us a "backward" function that accepts
(ctx, saved, *grads) and returns us the grad_inputs. The grad_inputs
are returned as a dict mapping str to a gradient.
Please see the changes in custom_op_db.py for examples of the API.

There are a number of pieces to this PR and I'm happy to split it if it
helps. They are:
- The actual APIs for specifying the two functions
(impl_save_for_backward, impl_backward)
- The autograd kernel: we take the functions the user give us and
construct an autograd.Function object that we then register to
the Autograd dispatch key
- Indirection for the autograd kernel. We add a layer of indirection so
that one can swap out the autograd kernel. This is necessary because by
default, we register an "autograd not implemented" kernel as the
Autograd implementation but then swap it for the actual kernel when the
user provides it.

Test Plan:
- We apply this API to give backward formulas for things in
custom_op_db. We then hook up custom_op_db to the Autograd OpInfo tests.
- Various tests in test_python_dispatch.py to check error cases.

zou3519 committed:
[custom_op] explicit autograd API

This PR adds an explicit API for registering a backward formula for a
CustomOp. In the end state, we will likely have this explicit API and a
magic API (which is sugar on top of an explicit API), since different
parties of users prefer different ones.

Concretely, to define a backward formula for a CustomOp:
- a user must provide us a "save for backward" function that accepts
(inputs, output) and returns exactly what they want saved for backward
- a user must provide us a "backward" function that accepts
(ctx, saved, *grads) and returns us the grad_inputs. The grad_inputs
are returned as a dict mapping str to a gradient.
Please see the changes in custom_op_db.py for examples of the API.

There are a number of pieces to this PR and I'm happy to split it if it
helps. They are:
- The actual APIs for specifying the two functions
(impl_save_for_backward, impl_backward)
- The autograd kernel: we take the functions the user give us and
construct an autograd.Function object that we then register to
the Autograd dispatch key
- Indirection for the autograd kernel. We add a layer of indirection so
that one can swap out the autograd kernel. This is necessary because by
default, we register an "autograd not implemented" kernel as the
Autograd implementation but then swap it for the actual kernel when the
user provides it.

Test Plan:
- We apply this API to give backward formulas for things in
custom_op_db. We then hook up custom_op_db to the Autograd OpInfo tests.
- Various tests in test_python_dispatch.py to check error cases.

[ghstack-poisoned]

ezyang commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+    assert isinstance(args, tuple)
+    attribs = [arg.name for arg in schema.arguments.flat_all]
+    name = str(schema.name) + "Args"
+    tuple_cls = namedtuple(name, attribs)
```
Do you want to cache this or something, so that repeated calls with the same schema return consistent named tuple objects?

albanD replied:
moving this to a helper with functools.cache() sounds good.

ezyang reviewed (approved):
LGTM. But perhaps you want to have one of the autograd experts take a look over.

ezyang reviewed (approved):
LGTM. But perhaps you want to have one of the autograd experts take a look over.

albanD commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+
+    expected_keys = {arg.name for arg in forward_op._schema.arguments.flat_all
+                     if arg.type.is_tensor_like()}
+    actual_keys = {k for k, _ in grad_inputs_dict.items()}
```
grad_inputs_dict.keys() ?
You can compare the result of this with the dict above or a set.

zou3519 replied:
I was a bit scared that the type of `dict.keys()` is not actually a dict or a set, but I'll try it out

zou3519 replied:
Yes, this works

albanD commented on torch/_custom_op/impl.py:
```diff
@@ -417,6 +415,87 @@ def error_on_ctx():
+
+        Use `impl_save_for_backward` to define a "save for backward" function
+        that specifies what gets saved for backward. The function should accept
+        two arguments ``(inputs, output)`` and return the quantities to be saved
```
inputs, outputs ?

zou3519 replied:
No, this is actually `inputs, output`:
- If the output is a single Tensor, then `output` is a single Tensor
- If the output is a Tuple of Tensors, then `output` is a Tuple of Tensors
- If the output is a tuple of a single Tensor, then `output` is a Tuple of a single Tensor
I'm open to changing it (though, this is how setup_context works as well).

albanD replied:
I'm curious:
- Why do you prefer this semantic?
- What is the semantic used in our other similar functions? (hooks always wrap in a tuple IIRC?)

zou3519 replied:
> Why do you prefer this semantic?

- It's easier to explain that the `output` is exactly the output of the `forward()`, rather than "we wrap the output in a Tuple if it is not a Tuple".
- It avoids ambiguity: If we support returning TensorList (which we likely will in the future, I have not enabled it in this PR), and we have `foo(Tensor x) -> TensorList`, then there is the question of if `outputs` is a Tuple[List[Tensor]] or just a List[Tensor].

> What is the semantic used in our other similar functions? (hooks always wrap in a tuple IIRC?)

- nn.Module `register_forward_hook` passes the [output as-is without wrapping](https://gist.github.com/zou3519/822505b737f8bc0646e45680bc7798cc)
- `register_backward_hook` (which requires hook of signature `(mod, grad_input, grad_output)` wraps grad_output as a Tuple.
- autograd.Function's forward distinguishes between `return (x,)` and `return x`
- autograd.Function's backward does not distinguish between `return grad` and `return (grad,)`

Based on those examples it seems like for the forward pass we are generally careful that the types match, but when it comes to gradients we generally wrap things in a Tuple to make things simpler.

albanD replied:
Ok! Keeping it as is sounds good then!

albanD commented on torch/_custom_op/impl.py:
```diff
@@ -417,6 +415,87 @@ def error_on_ctx():
+        The backward function must return a dict that maps the name of
+        an input to the CustomOp to its corresponding gradient. All inputs that
+        were declared to be Tensors in the CustomOp definition must be accounted
+        for in the dict. The gradient may be a Tensor or None.
```
I'm not sure what's the benefit of using a dict here vs a tuple where there should be a 1-1 match anymore?
Is it just so that we don't need to pad with Nones anymore?

In the original design we were using dicts here for the registration API, it was a nice way to specify what each function was computing. But if this is doing eager evaluation is it still needed?

zou3519 replied:
> I'm not sure what's the benefit of using a dict here vs a tuple where there should be a 1-1 match anymore?
Is it just so that we don't need to pad with Nones anymore?

The main motivation is so that we don't need to pad with Nones anymore. Padding with Nones was a bit annoying to do, especially with autograd.Function that had a long list of inputs. I can be convinced to go back to the 1:1 match.

> In the original design we were using dicts here for the registration API, it was a nice way to specify what each function was computing. But if this is doing eager evaluation is it still needed?

We can still have a separate derivatives.yaml-like API for this, since there are separate technical benefits to the separate derivatives.yaml declarations vs the autograd.Function style. That one can look something like:

```py
myop.def_autograd({
   'x': ...
   'y': ...
})

```

It may be confusing to use dicts in both places, though.

albanD replied:
> We can still have a separate derivatives.yaml-like API for this

I was not suggesting that no. I agree we shouldn't mix them.

> The main motivation is so that we don't need to pad with Nones anymore.

Sounds fair!
We can always add tuple support if some people prefer later on.

albanD commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -56,6 +66,22 @@ def numpy_mul_abstract(x, y):
+
+@numpy_mul.impl_backward()
+def numpy_mul_backward(ctx, saved, grad_out):
+    grad_x = grad_out * saved['y'] if saved['x_requires_grad'] else None
```
The ctx should be telling you what to compute gradients for really.
This is fine as a test but we should not recommend users do this. We should just expose `needs_input_grad` on the ctx.

zou3519 replied:
Agree, I didn't add needs_input_grad in this PR because it was getting long.

zou3519 replied:
I can add it as a follow-up, unless you want to see it in this PR?

albanD replied:
No need to add it here as long as this is not in any user-facing example!

albanD commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+            save_for_backward_fn_inputs = namedtuple_args(schema, args)
+            to_save = save_for_backward_fn(save_for_backward_fn_inputs, output)
+
+            save_pytree_for_backward(ctx, (to_save, args_info))
```
nit we should be able to skip this when grad mode is disabled?

zou3519 replied:
Sure

zou3519 replied:
It is a little complicated to do this because autograd.Function's forward sets grad_mode to False. This means that we would need to shepherd the knowledge of if grad mode is disabled or not from somewhere else.

Since this is a nit I'm just going to add this to the wishlist.

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+from collections import namedtuple
+
+
+# NOTE [CustomOp autograd kernel indirection]
```
This was meant to be the note, I didn't fill it out. I'll fill it out

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+    # or change the default autograd fallback to the autograd not implemented fallback.
+    def autograd_not_implemented(*args, **kwargs) -> None:
+        if pytree.tree_any(
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
```
Yes, good point

zou3519 replied:
It is a good idea to give people an easy way to do this (so they don't need to define their save_for_backward function and their backward function in order to say that their function is not differentiable). I can address that as a follow up, unless you want to see it in this PR?

I'm not really sure what this should look like yet: we could ask the author to give us a `differentiable` flag at creation of the custom_op, or we could ask them to `impl_backward(None)`, or something else.

soulitzer replied:
Follow up PR sounds ok!

> I'm not really sure what this should look like yet: we could ask the author to give us a differentiable flag at creation of the custom_op, or we could ask them to impl_backward(None), or something else.

My vote would be for having a `differentiable` flag at custom_op creation since whether the op is differentiable or not seems like a property of the op itself.

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+        if custom_op._has_impl('autograd'):
+            kernel = custom_op._get_impl('autograd').func
+            return kernel(*args, **kwargs)
+        if custom_op._has_impl('save_for_backward') or custom_op._has_impl('backward'):
```
That is the case, yes. I'll add a comment

zou3519 replied:
> Is it so that we can avoid passing save_for_backward through the decorator?

Yes. It felt strange that we were passing a function through a decorator, so I split it into two decorators.

> The argument for combining it is that since we always need both, might as well have them registered in the same place.

Yeah I agree that this is an argument for the opposite.

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
+        ):
+            raise RuntimeError("Autograd has not been implemented for operator")
+        guard = torch._C._AutoDispatchBelowAutograd()
```
https://github.com/pytorch/pytorch/issues/101191

soulitzer commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+            # There is nothing on the ctx object for now, it is just there so
+            # that we can add additional things in the future.
+            inner_ctx = object()
+            grad_inputs_dict = backward_fn(inner_ctx, saved, *grads)
```
Curious what the API would look like with the implicit autograd API, are there any bc-breaking things we should be worried about when we transition to that one?

zou3519 replied:
By "implicit" do you mean "magic"?

We can just name the magic API something different; then there are no BC-breaking changes. But I might be misunderstanding the question
```py
@myop.impl_backward_unified(output_differentiability=...)
def backward(grad, x):
    return grad * x.cos()
```

soulitzer replied:
Oh yeah thats what I meant. That plan sounds good.

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -0,0 +1,222 @@
+            # Output must be one or more Tensors, no TensorList (yet)
+            if output_differentiability is not None:
+                if isinstance(output, tuple):
+                    assert len(output_differentiability) == len(output)
```
Yeah, this is checked in `impl_backward` (and we have tests)

zou3519 reviewed (commented):
.

soulitzer commented on test/test_python_dispatch.py:
```diff
@@ -785,6 +785,271 @@ def foo_impl(x):
+            def foo_backward(ctx, saved, grad):
+                return {'xs': None}
+
+    def test_backward_output_differentiability_elements(self):
```
Renaming also sounds good!

zou3519 replied:
Renamed

zou3519 committed:
Update on "[custom_op] explicit autograd API"


This PR adds an explicit API for registering a backward formula for a
CustomOp. In the end state, we will likely have this explicit API and a
magic API (which is sugar on top of an explicit API), since different
parties of users prefer different ones.

Concretely, to define a backward formula for a CustomOp:
- a user must provide us a "save for backward" function that accepts
(inputs, output) and returns exactly what they want saved for backward
- a user must provide us a "backward" function that accepts
(ctx, saved, *grads) and returns us the grad_inputs. The grad_inputs
are returned as a dict mapping str to a gradient.
Please see the changes in custom_op_db.py for examples of the API.

There are a number of pieces to this PR and I'm happy to split it if it
helps. They are:
- The actual APIs for specifying the two functions
(impl_save_for_backward, impl_backward)
- The autograd kernel: we take the functions the user give us and
construct an autograd.Function object that we then register to
the Autograd dispatch key
- Indirection for the autograd kernel. We add a layer of indirection so
that one can swap out the autograd kernel. This is necessary because by
default, we register an "autograd not implemented" kernel as the
Autograd implementation but then swap it for the actual kernel when the
user provides it.

Test Plan:
- We apply this API to give backward formulas for things in
custom_op_db. We then hook up custom_op_db to the Autograd OpInfo tests.
- Various tests in test_python_dispatch.py to check error cases.

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] explicit autograd API"


This PR adds an explicit API for registering a backward formula for a
CustomOp. In the end state, we will likely have this explicit API and a
magic API (which is sugar on top of an explicit API), since different
parties of users prefer different ones.

Concretely, to define a backward formula for a CustomOp:
- a user must provide us a "save for backward" function that accepts
(inputs, output) and returns exactly what they want saved for backward
- a user must provide us a "backward" function that accepts
(ctx, saved, *grads) and returns us the grad_inputs. The grad_inputs
are returned as a dict mapping str to a gradient.
Please see the changes in custom_op_db.py for examples of the API.

There are a number of pieces to this PR and I'm happy to split it if it
helps. They are:
- The actual APIs for specifying the two functions
(impl_save_for_backward, impl_backward)
- The autograd kernel: we take the functions the user give us and
construct an autograd.Function object that we then register to
the Autograd dispatch key
- Indirection for the autograd kernel. We add a layer of indirection so
that one can swap out the autograd kernel. This is necessary because by
default, we register an "autograd not implemented" kernel as the
Autograd implementation but then swap it for the actual kernel when the
user provides it.

Test Plan:
- We apply this API to give backward formulas for things in
custom_op_db. We then hook up custom_op_db to the Autograd OpInfo tests.
- Various tests in test_python_dispatch.py to check error cases.

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] explicit autograd API"


This PR adds an explicit API for registering a backward formula for a
CustomOp. In the end state, we will likely have this explicit API and a
magic API (which is sugar on top of an explicit API), since different
parties of users prefer different ones.

Concretely, to define a backward formula for a CustomOp:
- a user must provide us a "save for backward" function that accepts
(inputs, output) and returns exactly what they want saved for backward
- a user must provide us a "backward" function that accepts
(ctx, saved, *grads) and returns us the grad_inputs. The grad_inputs
are returned as a dict mapping str to a gradient.
Please see the changes in custom_op_db.py for examples of the API.

There are a number of pieces to this PR and I'm happy to split it if it
helps. They are:
- The actual APIs for specifying the two functions
(impl_save_for_backward, impl_backward)
- The autograd kernel: we take the functions the user give us and
construct an autograd.Function object that we then register to
the Autograd dispatch key
- Indirection for the autograd kernel. We add a layer of indirection so
that one can swap out the autograd kernel. This is necessary because by
default, we register an "autograd not implemented" kernel as the
Autograd implementation but then swap it for the actual kernel when the
user provides it.

Test Plan:
- We apply this API to give backward formulas for things in
custom_op_db. We then hook up custom_op_db to the Autograd OpInfo tests.
- Various tests in test_python_dispatch.py to check error cases.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented on torch/_custom_op/impl.py:
```diff
@@ -417,6 +415,90 @@ def error_on_ctx():
+        were declared to be Tensors in the CustomOp definition must be accounted
+        for in the dict. The gradient may be a Tensor or None.
+
+        TODO(rzou): Add example when this PR is closer to landing.
```
Sorry, I remembered this after I typed the merge command. Will do as a follow-up.

vadimkantorov commented:
Regarding the syntax, could `custom_op` be combined somehow with forward impl?
currently:
```python
        @custom_op(f'{TestCustomOp.test_ns}::foo')
        def foo(x: torch.Tensor) -> torch.Tensor:
            ...

        @foo.impl(['cpu', 'cuda'])
        def foo_impl(x):
            return x.sin()

        @foo.impl_backward()
        def foo_backward(ctx, saved, grad):
            return grad * saved.cos()
```

proposed:
```python
        @custom_op_with_forward(f'{TestCustomOp.test_ns}::foo', device_types = ['cpu', 'cuda'])
        def foo(x : torch.Tensor) -> torch.Tensor:
            return x.sin()

        @foo.impl_backward()
        def foo_backward(ctx, saved, grad):
            return grad * saved.cos()
```
